### DISCUSSION TITLE
请问如何查看中间生成的 llvm_ir ？

### DISCUSSION THREAD

**Discussion by nobodxbodon** [yuyan-lang/yuyan#73] [Q&A] [answered]
比如想查看 yy 和 yy_bs 生成的 llvm_ir 的差别。

**Comment by UltimatePea** [comment, marked as answer]
./.yybuild.nosync 目录下有所有的文件

yy 会生成一个.ll文件，名字是yy<一串数字>.ll
yy_bs 会生成很多.ll文件，名字不是yy<一串数字>.ll的文件都是yy_bs生成的

**Reply by nobodxbodon** [reply]
刚编译quick-sort.yuyan，yy_bs生成的文件中，只看到一个ll：
<img width="442" alt="Screenshot 2024-01-18 at 5 48 26 PM" src="https://github.com/yuyan-lang/yuyan/assets/392497/adf770ab-ebf2-474b-b3ca-6c134ea0f575">
另外，在 yy 生成的 yy30583.ll 中，等号左边的是类似于`v1574`的助记符，而在上面yy_bs生成的 `yy_quick-sort_豫言编译器默认入口文件.ll` 中，是类似于 `%"元组数组_entryMain_合并点227_重名_1535"`，是新版本改进了可读性吗？

**Reply by UltimatePea** [reply]
这个文件夹的子文件夹（比如.yybuild.nosync/yylib）里也有很多ll文件，都是编译生成的。

ll文件是给llvm读的，不是写给人读的，所以名称并没有太大意义。名字不同是因为新版本和旧版本采用了不同的标识符的生成方法。

**Reply by nobodxbodon** [reply]
IR 本身 [看来是多用途的](https://llvm.org/docs/LangRef.html#introduction)：
> The LLVM code representation is designed to be used in three different forms: as an in-memory compiler IR, as an on-disk bitcode representation (suitable for fast loading by a Just-In-Time compiler), ***and as a human readable*** assembly language representation.

不过上万行看起来的确很有些挑战。请问有没有什么阅读的建议？比如文件结构、源码逻辑部分集中在哪个文件等等？

**Reply by UltimatePea** [reply]
当然LLVM IR 比起 ELF字节码当然好读很多，但比起源代码还是要难读很多的。

**Reply by UltimatePea** [reply]
> 如文件结构、源码逻辑部分集中在哪个文件等等？

如果是编译器源码的话，src下是老版编译器所有源代码，`./豫言编译器`下是新版编译器所有源代码。

**Reply by UltimatePea** [reply]
不过目前老版编译器的开发应该完全停止了，新版编译器尚未开发完成，预计还会有挺大的改动。

**Reply by nobodxbodon** [reply]
> 不过上万行看起来的确很有些挑战。请问有没有什么阅读的建议？比如文件结构、源码逻辑部分集中在哪个文件等等？

额我的意思是读 ll 文件里的 IR 代码。

**Reply by UltimatePea** [reply]
这个嘛，我也没系统地读过编译器生成的LLVM IR的代码，没啥经验，如果你能够成功读懂的话，欢迎分享！

**Reply by nobodxbodon** [reply]
刚试了以下代码 `单元测试/ll/加.yuyan`：
```
寻观标准库之数据结构之整数操作之书。

「加」使用于「一」使用于「二」。
```
（建议在教程中先介绍一下如何使用标准库，否则大部分示例代码恐怕无法编译？）

将 二 改为 三 后，入口.ll （300行左右）相同，而 `.yybuild.nosync/单元测试/ll/加.ll` （~400行）中，仅第343行的 2 变 3：
`%"变_137_更新__提升函数106_合并点141437" = inttoptr i64 2 to i64*`

将 加 改为 减 后，以上 ll 中，第 60 和 249 行的 8 变为 15：
```
%"目标指针_豫言模块__Users_xuanwu_work_各种语言_yuyan_0_1_0alpha2_0_2_单元测试_ll_加_yuyan_计算函数_顶层续延名150_重名_221" = getelementptr i64*, i64** %"元组数组_豫言模块__Users_xuanwu_work_各种语言_yuyan_0_1_0alpha2_0_2_单元测试_ll_加_yuyan_计算函数_顶层续延名150_重名_221", i64 8
=> 
%"目标指针_豫言模块__Users_xuanwu_work_各种语言_yuyan_0_1_0alpha2_0_2_单元测试_ll_加_yuyan_计算函数_顶层续延名150_重名_221" = getelementptr i64*, i64** %"元组数组_豫言模块__Users_xuanwu_work_各种语言_yuyan_0_1_0alpha2_0_2_单元测试_ll_加_yuyan_计算函数_顶层续延名150_重名_221", i64 15
```

第 249 行：
```
%"目标指针_提升函数106_顶层续延名162_重名_364" = getelementptr i64*, i64** %"元组数组_提升函数106_顶层续延名162_重名_364", i64 8
=>
%"目标指针_提升函数106_顶层续延名162_重名_364" = getelementptr i64*, i64** %"元组数组_提升函数106_顶层续延名162_重名_364", i64 15
```

如改为 乘，则 8 变为 11，尚不明所以。

**Reply by UltimatePea** [reply]
（建议在教程中先介绍一下如何使用标准库，否则大部分示例代码恐怕无法编译？）

感谢建议！

目前 `寻观标准库之书。`就可以使用标准库

之后的话应该会把标准库直接写进语言底层，这样的话就是这一行也不用写，但现在还没到那一步。目前豫言仍然处于早期研发阶段，很多东西都可能会改，所以教程编写方面会比较谨慎。

**Reply by UltimatePea** [reply]
> 如改为 乘，则 8 变为 11，尚不明所以。

这个数字应该是这个操作在整数操作文件中是第几个函数。

**Reply by nobodxbodon** [reply]
> 这个数字应该是这个操作在整数操作文件中是第几个函数

哦谢谢指点。

如果整个 `导入并打开标准库。`，入口.ll 就有 2500多行，加.ll 也 440 多，还是导入单个库了。话说例程里没找到指定导入某个或某几个函数，不知支持吗？

刚在源码里加了赋值，改为： `结果的定义是「加」使用于「一」使用于「三」`。

对比赋值之前的 ll，标识符序号有变如第8行 _180 变为 _181：`%"元组数组_豫言模块__Users_xuanwu_work_各种语言_yuyan_0_1_0alpha2_0_2_单元测试_ll_加_yuyan_计算函数_入口续延名_重名_180" = bitcast i64* %"stack_top" to i64**`

如果要对比两个 ll，是否可作个遍历，尽量找到所有 _180->_181 这种映射之后替换，再过滤所有相同语句？

刚想试加上条件，只找到了三段的：
```
如果五等于零
    那么「加」使用于「一」使用于「三」
    否则「加」使用于「二」使用于「四」 。
```
若去掉“否则”从句就解析报错，不知是否支持类似“如果〇那么〇”这样的单条件语句？

**Reply by UltimatePea** [reply]
> 话说例程里没找到指定导入某个或某几个函数，不知支持吗？

不支持，但可以导入不打开：

导入标准库。
「加」者标准库之。。。也。

> 若去掉“否则”从句就解析报错，不知是否支持类似“如果〇那么〇”这样的单条件语句？

不支持，我印象中没有函数式语言支持单条件从句的

**Reply by UltimatePea** [reply]
但是可以用模式匹配实现相同的效果，会抛出编译时（以后的版本）或运行时的错误（现在的版本）。

**Reply by UltimatePea** [reply]
> 如果要对比两个 ll，是否可作个遍历，尽量找到所有 _180->_181 这种映射之后替换，再过滤所有相同语句？

这种末尾的数字都是随机生成的，微小的改动都会使数字发生变化。我不太清楚具体你要用ll文件干什么，但你可以尝试各种方法。

**Reply by UltimatePea** [reply]
如果你真的想研究LLVM源代码的话

在全局配置里把

字节码打印调试信息乃引用类于（爻）也。
字节码打印调试信息者新引用于（阴）也。

的值改成阳，重新编译yy_bs，然后在编译文件的时候用 `--parallel --debug --save-build-artifact`，
把`*.代码生成前形式.json`的文件内容和LLVM IR的内容对照起来看

json文件可以用 `./yy_bs debug showtrees <json路径>` 来打印

**Reply by nobodxbodon** [reply]
把 ` 字节码打印调试信息者新引用于（阴）也。` 的阴改为了阳后 make yy_bs 通过。

是这样编译吗？
```
% ./yy_bs --parallel --debug --save-build-artifact 单元测试/ll/加.yuyan
！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！
豫言运行环境(yy_runtime)：未捕捉的异常(Uncaught Exception)：
尝试打印值：（可能会出现segfault）：
不识别的命令行选项：--save-build-artifact
```

另外，之前生成的这个json成功树形打印了：
```
% ./yy_bs debug showtrees .yybuild.nosync/单元测试/ll/加.求值正则形式.json
豫言模块__Users_xuanwu_work_各种语言_yuyan_0_1_0alpha2_0_2_单元测试_ll_加_yuyan_计算函数（阳）的定义是：λ全局函数空参数一105.
    虑 未初始化元素107  者  (空)[]  而
...
```
但：
```
% ./yy_bs debug showtrees .yybuild.nosync/单元测试/ll/加.抽象语法.json
！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！
豫言运行环境(yy_runtime)：未捕捉的异常(Uncaught Exception)：
尝试打印值：（可能会出现segfault）：
无法解析癸象(742)：{"类型":"式节点","节点名称":{"名称":"近似结构节点","标签名":null},"参数":[{"类型":"式节点","节点名称":{"名称":"文件导入近似类型节点","串":"/Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/yylib/标准库/数据结构/整数操作。豫"},"参数":[]},{"类型":"绑定","可能名":"整数操作","绑定体":{"类型":"式节点","节点名称":{"名称":"近似结构节点","标签名":null},"参数":[{"类型":"式节点","节点名称":{"名称":"定义声明近似类型节点","透明性":"透明"},"参数":[{"类型":"式节点","节点名称":{"名称":"派抽象","参式":"显式"},"参数":[{"类型":"式节点","节点名称":"近似类型原子节点","参数":[]},{"类型":"绑定","可能名":null,"绑定体":{"类型":"式节点","节点名称":{"名称":"派抽象","参式":"显式"},"参数":[{"类型":"式节点","节点名称":"近似类型原子节点","参数":[]},{"类型":"绑定","可能名":null,"绑定体":{"类型":"式节点","节点名称":"近似类型原子节点","参数":[]}}]}}]},{"类型":"式节点","节点名称":{"名称":"结构投影","串":"加"},"参数":[1]}]},{"类型":"绑定","可能名":"加","绑定体":{"类型":"式节点","节点名称":{"名称":"近似结构节点","标签名":"结果"},"参数":[{"类型":"式节点","节点名称":{"名称":"定义声明近似类型节点","透明性":"不透明"},"参数":[{"类型":"式节点","节点名称":"近似类型原子节点","参数":[]},{"类型":"式节点","节点名称":{"名称":"函数调用","参式":"显式"},"参数":[{"类型":"式节点","节点名称":{"名称":"函数调用","参式":"显式"},"参数":[1,"一"]},"三"]}]},{"类型":"绑定","可能名":"结果","绑定体":{"类型":"式节点","节点名称":"空近似结构节点","参数":[]}}]}}]}}]}
```

**Reply by UltimatePea** [reply]
> 不识别的命令行选项：--save-build-artifact


这个东西要等 #70 merge了才能用，目前这个branch还有一点小bug，可以先不加--save-build-artifact，但会少信息，

用的话可能这个commit 应该是没问题的 bd920c27621e96a815c80eb526bb9288952292a5，但其他的commit可能有编译bug

有两个，一个是showtrees， 一个是 showtree，不同的json要用不同的
% ./yy_bs debug showtree .yybuild.nosync/单元测试/ll/加.抽象语法.json

**Reply by nobodxbodon** [reply]
--parallel 好像报错，这参数是何用？
```
'======================================='
Waiting for updates...
exception calling callback for <Future at 0x7fe7c03ac070 state=finished raised ValueError>
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/xuanwu/opt/anaconda3/lib/python3.8/concurrent/futures/process.py", line 239, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/parallel_compile.py", line 191, in worker
    process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, preexec_fn=pre_fun(stages.index(stage)))
ValueError: 'type-check-and-erase' is not in list
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/xuanwu/opt/anaconda3/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
    callback(self)
  File "parallel_compile.py", line 324, in process_result
    (result, error) = future.result()
  File "/Users/xuanwu/opt/anaconda3/lib/python3.8/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/Users/xuanwu/opt/anaconda3/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
ValueError: 'type-check-and-erase' is not in list
completed type-check /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/单元测试/ll/加.yuyan
'======================================='
'======================================='
'============== Scheduled =============='
{'dependency-analysis': 0,
 'parse': 0,
...
```

showtree 输出是像这样吗？
```
 % ./yy_bs debug showtree .yybuild.nosync/单元测试/ll/加.抽象语法.json

《《《？Σ⟨?⟩(↪_)[近似文件导入(/Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/yylib/标准库/数据结构/整数操作。豫)[]；整数操作.Σ⟨?⟩(↪_)[透明近似定义声明[Π[(*)[]；_.Π[(*)[]；_.(*)[]]]；↘加↙[1]]；加.Σ⟨?⟩(↪结果)[不透明近似定义声明[(*)[]；ap[ap[1；「 一 」]；「 三 」]]；结果.Σ⟨?⟩0[]]]]》》》。。。。
」
```

**Reply by UltimatePea** [reply]
> --parallel 好像报错，这参数是何用？

这个是并行编译，可能有bug，我得找时间看一下原因。我的并行编译过程会生成很多的中间json文件，方便查看各个步骤的语法，非并行编译的--save-build-artifact还没写

> showtree 输出是像这样吗？

是这样的，但最新版本的yy_bs更新了显示方式
